File: apps/backend/functions/test/contract/booking_semantic_hash.test.ts
import assert from 'node:assert/strict';
import { test } from 'node:test';

import {
  asRecord,
  createBookingSchemaRegistry,
  readBookingJson,
} from '../support/booking_schema_registry.js';
import { StrictJsonReader } from '../support/booking_raw_json.js';
import { computeBookingSemanticHash } from '../support/booking_semantic_hash.js';

void test('shared vectors freeze independent TypeScript JCS bytes and SHA-256', async () => {
  const registry = await createBookingSchemaRegistry();
  const validateCommand = registry.validator('booking_command.schema.json');
  const container = asRecord(await readBookingJson('fixtures/semantic_hash_vectors.json'));
  assert.ok(Array.isArray(container.vectors));
  const results = new Map<string, ReturnType<typeof computeBookingSemanticHash>>();

  for (const raw of container.vectors) {
    const vector = asRecord(raw);
    const rawCommand = String(vector.rawCommandJson);
    assert.equal(validateCommand(new StrictJsonReader(rawCommand).read()), true, String(vector.id));
    const result = computeBookingSemanticHash(
      rawCommand,
      asRecord(vector.resolvedActorScope),
      validateCommand,
    );
    assert.deepEqual(result.projection, vector.expectedProjection, String(vector.id));
    assert.equal(result.canonicalHex, vector.expectedCanonicalHex, String(vector.id));
    assert.equal(result.digest, vector.expectedDigest, String(vector.id));
    results.set(String(vector.id), result);
  }

  assert.equal(
    results.get('key-order-a')!.digest,
    results.get('key-order-b-request-change')!.digest,
  );
  assert.equal(
    results.get('nested-unicode-nfc-literal')!.digest,
    results.get('nested-unicode-nfc-escaped')!.digest,
  );
  assert.notEqual(
    results.get('nested-unicode-nfc-literal')!.digest,
    results.get('unicode-nfd-distinct')!.digest,
  );
  assert.notEqual(
    results.get('booking-revision-7')!.digest,
    results.get('booking-revision-8')!.digest,
  );
  assert.notEqual(
    results.get('event-revision-11')!.digest,
    results.get('event-revision-12')!.digest,
  );
  assert.equal(results.get('idempotency-a')!.digest, results.get('idempotency-b')!.digest);
  assert.notEqual(
    results.get('idempotency-a')!.logicalIdentity,
    results.get('idempotency-b')!.logicalIdentity,
  );
});

void test('raw invalid vectors are rejected before hashing', async () => {
  const registry = await createBookingSchemaRegistry();
  const validateCommand = registry.validator('booking_command.schema.json');
  const container = asRecord(await readBookingJson('fixtures/semantic_hash_invalid.json'));
  assert.ok(Array.isArray(container.cases));
  for (const raw of container.cases) {
    const fixture = asRecord(raw);
    assert.throws(
      () =>
        computeBookingSemanticHash(
          String(fixture.rawCommandJson),
          asRecord(fixture.resolvedActorScope),
          validateCommand,
        ),
      /.+/u,
      `${String(fixture.id)}: ${String(fixture.reason)}`,
    );
  }
});

File: packages/api_contracts/test/booking_semantic_hash_test.dart
import 'package:test/test.dart';

import 'support/booking_schema_fixture_validator.dart';
import 'support/booking_semantic_hash.dart';

void main() {
  final vectors = readJsonObject(
    '$bookingSchemaRoot/fixtures/semantic_hash_vectors.json',
  );
  final invalid = readJsonObject(
    '$bookingSchemaRoot/fixtures/semantic_hash_invalid.json',
  );

  test('shared vectors freeze projection, canonical bytes and SHA-256', () {
    final results = <String, BookingSemanticHashResult>{};
    for (final vector in readObjectList(vectors, 'vectors')) {
      final result = computeBookingSemanticHash(
        rawCommandJson: vector['rawCommandJson']! as String,
        resolvedActorScope: (vector['resolvedActorScope']! as Map)
            .cast<String, Object?>(),
      );
      expect(
        result.projection,
        vector['expectedProjection'],
        reason: vector['id']! as String,
      );
      expect(
        result.canonicalHex,
        vector['expectedCanonicalHex'],
        reason: vector['id']! as String,
      );
      expect(
        result.digest,
        vector['expectedDigest'],
        reason: vector['id']! as String,
      );
      results[vector['id']! as String] = result;
    }

    expect(
      results['key-order-a']!.digest,
      results['key-order-b-request-change']!.digest,
    );
    expect(
      results['nested-unicode-nfc-literal']!.digest,
      results['nested-unicode-nfc-escaped']!.digest,
    );
    expect(
      results['nested-unicode-nfc-literal']!.digest,
      isNot(results['unicode-nfd-distinct']!.digest),
    );
    expect(
      results['booking-revision-7']!.digest,
      isNot(results['booking-revision-8']!.digest),
    );
    expect(
      results['event-revision-11']!.digest,
      isNot(results['event-revision-12']!.digest),
    );
    expect(results['idempotency-a']!.digest, results['idempotency-b']!.digest);
    expect(
      results['idempotency-a']!.logicalIdentity,
      isNot(results['idempotency-b']!.logicalIdentity),
    );
  });

  test('raw invalid vectors fail before hashing', () {
    for (final fixture in readObjectList(invalid, 'cases')) {
      expect(
        () => computeBookingSemanticHash(
          rawCommandJson: fixture['rawCommandJson']! as String,
          resolvedActorScope: (fixture['resolvedActorScope']! as Map)
              .cast<String, Object?>(),
        ),
        throwsFormatException,
        reason: '${fixture['id']}: ${fixture['reason']}',
      );
    }
  });
}
